src/schema/validate.rs

```diff
@@ -0,0 +1,1001 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+
+#![allow(dead_code)]
+
+use crate::{
+    schema::{Schema, Type},
+    *,
+};
+use alloc::collections::BTreeMap;
+use regex::Regex;
+
+type String = Rc<str>;
+
+/// Validation errors that can occur when validating a Value against a Schema.
+#[derive(Debug, Clone, PartialEq, Eq)]
+pub enum ValidationError {
```

Comment:
>ValidationError [](http://example.com/codeflow?start=9&length=15)

is this normally the pattern in rust to put multiple entities in one file?

Reply:
In Rust, each file is a module and a module can be specified only in a single file.
Sub modules can be specified in their own separate files within a directory whose name matches the parent module.

I have not seen too much of a standard pattern here. There is no hard rule that each entity must exist in its own file.
It's more about the file size I guess. In this case, we could actually move the ValidationError to its own separate file (module) and refer to it elsewhere.